src/Parser/VariadicMethodsVisitor.php

```diff
@@ -0,0 +1,108 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Parser;
+
+use PhpParser\Node;
+use PhpParser\Node\Name;
+use PhpParser\Node\Stmt\ClassMethod;
+use PhpParser\NodeVisitorAbstract;
+use PHPStan\Reflection\ParametersAcceptor;
+use PHPStan\TrinaryLogic;
+use function array_key_exists;
+use function in_array;
+
+final class VariadicMethodsVisitor extends NodeVisitorAbstract
+{
+
+	private ?Node $topNode = null;
+
+	private ?string $inNamespace = null;
+
+	private ?string $inClassLike = null;
+
+	private ?string $inMethod = null;
+
+	/** @var array<string, array<string, TrinaryLogic>> */
+	private array $variadicMethods = [];
+
+	public const ATTRIBUTE_NAME = 'variadicMethods';
+
+	public function beforeTraverse(array $nodes): ?array
+	{
+		$this->topNode = null;
+		$this->variadicMethods = [];
+		$this->inNamespace = null;
+		$this->inClassLike = null;
+		$this->inMethod = null;
+
+		return null;
+	}
+
+	public function enterNode(Node $node): ?Node
+	{
+		if ($this->topNode === null) {
+			$this->topNode = $node;
+		}
+
+		if ($node instanceof Node\Stmt\Namespace_ && $node->name !== null) {
+			$this->inNamespace = $node->name->toString();
+		}
+
+		if ($node instanceof Node\Stmt\ClassLike && $node->name instanceof Node\Identifier) {
+			$this->inClassLike = $this->inNamespace !== null ? $this->inNamespace . '\\' . $node->name->name : $node->name->name;
```

Comment:
This will break on anonymous classes. Please add some tests that would get thrown off by this logic, and then fix it by having an array stack instead of a single property value.

Reply:
hmm. 

to properly support anonymous classes I would have to create a name key which I can later on reproduce in `PhpMethodReflection` so I can find the correct array-offsets again in https://github.com/phpstan/phpstan-src/pull/3534/files#diff-aa8ffb29754c75aa88928c72d3877f0f030a934d9cff263dfab527aaf4037172R256-R260

the name returned by `ClassReflection->getName()` [depends on the filename](https://github.com/phpstan/phpstan-src/blob/e2654b7dc87951e1481bc46d8f1f7ff587e3f484/src/Broker/AnonymousClassNameHelper.php#L23-L48). In this Visitor I don't have access to the file-name though.

when building a different key e.g. based on the AST-Node doesn't work either because in `PhpMethodReflection` I don't have access to the AST.

Reply:
It's fine if we don't find variadic methods inside anonymous classes (I don't think we do now either), it's sufficient if the code doesn't break on classes containing anonymous classes.

Reply:
is this test good enough?

https://github.com/phpstan/phpstan-src/pull/3534/files#diff-c2d7f467d27d2e9f8c672d41471a89079412f952f7e686ef9d3fc7a431c8e782R15-R21

Reply:
ohh I see the problem now

Reply:
We need a test that breaks the visitor.

Use-case: There's anonymous class in the middle of a class. After the visitor leaves the anonymous class, the class name is set to null. Which means that if there's a class method with `func_get_args()` call, the visitor is not going to correctly set the attribute for it.